### PR TITLE
Improvements related to `databake::test_bake!()`

### PR DESCRIPTION
**Pull request by alvra**
This PR improves the usefulness of the `test_bake!` macro by making sure the baked tokenstream matches the literal constructor.

- update the implementations for slices and tuples 
- convert corresponding tests to use `test_bake!`
- enable some tests that now pass
- more test coverage for slices and Cow

I've also tried to modify the output of the `Bake` derive macro so `test_bake!` works everywhere but I couldn't figure out the logic for when commas are omitted or not so that's not part of this PR. Unfortunately this means that derived implementations still cannot always be tested using `test_bake!`. But with this PR, at least primitives, manual implementations, and combinations thereof will now work.

### CODE REVIEW

**Review by robertbastian — COMMENTED** [review]
Thanks, this is great! Did you check whether any other mentions of `https://github.com/rust-lang/rust/issues/98906` in the repo can be reenabled now?

**Review by sffc — APPROVED** [review]
Does this also fix the other instances of https://github.com/rust-lang/rust/issues/98906 scattered through the codebase?

File: utils/databake/src/alloc.rs [hunk 1]
```diff
@@ -25,6 +25,13 @@ fn cow() {
         const: alloc::borrow::Cow::Borrowed("hi"),
         alloc
     );
+    assert_eq!(
+        Bake::bake(
+            &alloc::borrow::Cow::<'static, u8>::Owned(1),
+            &Default::default(),
+        ).to_string(),
+        "alloc :: borrow :: Cow :: Borrowed (& 1u8)",
```
[the remark this answers is not in the document]
Comment (alvra) [reply utils/databake/src/alloc.rs:33]: I've added this check to the test.

File: utils/databake/src/primitives.rs [hunk 2]
```diff
@@ -193,6 +206,16 @@ fn result() {
 }
 
 macro_rules! tuple {
+    ($ty:ident, $ident:ident) => {
+        impl<$ty> Bake for ($ty,) where $ty: Bake {
```
Comment (Manishearth) [inline utils/databake/src/primitives.rs:210]: nit: can we just redirect to the same macro with a comma tacked on

**Comment by alvra** [comment]
I've enables 2 tests outside of `databake` that now pass.
- For one of these I had to update the test itself too, could you verify this is correct?
- There's another test ignored because of `rust#98906` that fails for a different reason and could probably be enabled: `components/list/src/provider/serde_dfa.rs:L231`

The other 3 tests ignored use de98906rived `Bake` impls so they still fail over commas unfortunately.

**Comment by Manishearth** [comment]
This is great, thanks!

File: utils/databake/src/primitives.rs (hunk 2 above)
Comment (alvra) [reply utils/databake/src/primitives.rs:210]: Not sure how to do that without it being too convoluted. What did you have in mind?

Comment (Manishearth) [reply utils/databake/src/primitives.rs:210]: wait never mind I misunderstood this macro, I thought it was recursive

File: utils/databake/src/alloc.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (robertbastian) [reply utils/databake/src/alloc.rs:33]: I'm not a big fan of testing the exact string. It's enough to test that the borrowed variants is a bake fix-point, and that the owned variant has the same bake as the borrowed variant, regardless of what the actual string representation is. Consider removing this check.

[the remark this answers is not in the document]
Comment (alvra) [reply utils/databake/src/alloc.rs]: Agreed

**Review by robertbastian — APPROVED** [review]

**Pull request merged**